### PR TITLE
podspec support for builders in env

### PR DESCRIPTION
**Pull request by viveksinghggits**
this blog explains the fix that we did in 
https://github.com/fission/fission/pull/1369

### CODE REVIEW

File: content/posts/fission_builder_podspec.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  poolsize: 3
+  resources: {}
+  runtime:
+    image: fission/python-env
+  terminationGracePeriod: 360
+  version: 2
+{{< /highlight  >}}
+
+ 
+We can see that for builder we are just specifying the command and image, and if we look into the pod that will be created, if we apply this spec, it will only have standard podspec that will be set by fission itself.
```
Comment (bhavin192) [inline content/posts/fission_builder_podspec.md]: Fission instead of fission

File: content/posts/fission_builder_podspec.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  keeparchive: false
+  poolsize: 3
+  resources: {}
+  runtime:
+    image: fission/python-env
+    terminationGracePeriod: 360
+  version: 2
+{{< /highlight  >}}
+
+You can see that for builder we are now specifying the podspec and in that we can have almost everything that is supported in kuberentes podspec specification, for example to satisfy our requirement to schedule the builder pod only on the nodes that have label `machinecap=high` we have introduced [nodeSelector](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector) in the builder podspec that takes care of this requirement.
```
Comment (bhavin192) [inline content/posts/fission_builder_podspec.md]: s/kuberentes podspec specification/Kubernetes PodSpec specification/ 
Might want to add link to API doc here? https://v1-16.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#podspec-v1-core

File: content/posts/fission_builder_podspec.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    key: node.kubernetes.io/not-ready
+    operator: Exists
+    tolerationSeconds: 300
+  - effect: NoExecute
+    key: node.kubernetes.io/unreachable
+    operator: Exists
+    tolerationSeconds: 300
+{{< /highlight >}}
+
+I am running this example on a cluster that is provisioned through minikube and if I check the status of that pod we see that the pod is in `pending` state
```
Comment (bhavin192) [inline content/posts/fission_builder_podspec.md]: s/`pending`/`Pending`

File: content/posts/fission_builder_podspec.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+```
+# kubectl describe pod -n fission-builder py-builder-1971419-6dd5658569-c88tc
+
+Events:
+  Type     Reason            Age              From               Message
+  ----     ------            ----             ----               -------
+  Warning  FailedScheduling  7s (x5 over 6m)  default-scheduler  0/1 nodes are available: 1 node(s) didn't match node selector.
+
+```
+you might have guessed that this is becasue we have mentioned the `nodeSelector` in the environment spec but we dont have any node in the cluster that has this label alreaedy set. If we want this pod to be running on the minikube node that we have, we will have to set the label in the minikube node, lets try to do that and then check the status
```
Comment (bhavin192) [inline content/posts/fission_builder_podspec.md]: s/dont/don't
s/lets/let's

File: content/posts/fission_builder_podspec.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+```
+# kubectl get pod -n fission-builder   
+NAME                                  READY   STATUS    RESTARTS   AGE
+py-builder-1974124-865f5fd498-2m9sn   2/3     Running   0          16s
+```
+Describing the pod will let you know that the pod now has been created with the correct tolerations that we have specified for builder in the environment spec.
+
+**_Author:_**
+
+* [Vivek Singh](https://viveksingh.dev)  **|**  Software Engineer - [Infracloud Technologies](http://infracloud.io/)
```
Comment (bhavin192) [inline content/posts/fission_builder_podspec.md]: s/Infracloud/InfraCloud

File: content/posts/fission_builder_podspec.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+# kubectl get pods -n fission-builder 
+NAME                                  READY   STATUS    RESTARTS   AGE
+py-builder-1971419-6dd5658569-c88tc   2/3     Running   0          10m
+```
+
+So, as you can see in the example given we can now provide `podSpec` for builder while creating an environment, which enables us a lot of new functionality. 
+
+## Using tolerations in builder podspec 
+Lets take another example, we will now taint the node minikube and then try to provide toleration in the pods to get it successfully scheduled on the minikube node. Set the tain on the node using below command
```
Comment (bhavin192) [inline content/posts/fission_builder_podspec.md]: s/Lets/Let's

File: content/posts/fission_builder_podspec.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  keeparchive: false
+  poolsize: 3
+  resources: {}
+  runtime:
+    image: fission/python-env
+    terminationGracePeriod: 360
+  version: 2
+{{< /highlight >}}
+
+You can see that we are setting `tolerations` for builder pod in the environment spec. Lets apply this environment spec and check if the pod is scheduled on `minikube` node successfully.
```
Comment (bhavin192) [inline content/posts/fission_builder_podspec.md]: s/Lets/Let's

**Comment by life1347** [comment]
Overall, it looks good to me. I'm thinking it would be great if you can also add a simple spec example to https://github.com/fission/fission/tree/master/examples/spec-example so that readers can follow the instruction to do the tests themselves.

**Comment by life1347** [comment]
It there is no further change, I will merge it.

**Comment by viveksinghggits** [comment]
Hi @life1347,
Since the `spec-example` directory contains the example of how we can use spec to create the `fission resources` I am thinking to have another directory in `examples` named `podspec-example`, which will contain the pod-spec example.

**Comment by life1347** [comment]
Since podspec is also part of spec usage, we can create a new directory called `hello-spec-exmaple` and move all existing files to it. Then, you can create a `podspec-example` directory at the same level of `hello-spec-exmaple`.

**Comment by viveksinghggits** [comment]
Hi @life1347, 
Raised  https://github.com/fission/fission/pull/1374 for adding examples.

File: content/posts/fission_builder_podspec.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+{{< /highlight >}}
+
+You can see that we are setting `tolerations` for builder pod in the environment spec. Let's apply this environment spec and check if the pod is scheduled on `minikube` node successfully.
+```
+# kubectl get pod -n fission-builder   
+NAME                                  READY   STATUS    RESTARTS   AGE
+py-builder-1974124-865f5fd498-2m9sn   2/3     Running   0          16s
+```
+Describing the pod will let you know that the pod now has been created with the correct tolerations that we have specified for builder in the environment spec.
+
```
Comment (life1347) [inline content/posts/fission_builder_podspec.md]: How about we add a link to the example podspec? That would be very helpful for users.

Comment (viveksinghggits) [reply content/posts/fission_builder_podspec.md]: will we have to wait for example `podspec` to be merged?

Comment (life1347) [reply content/posts/fission_builder_podspec.md]: ok, let's merge it first.

**Pull request merged**